### PR TITLE
call _transitionEnd if no transition is performed, styles need to be set

### PR DESCRIPTION
**Pull request by beta-vulgaris**
This fixes the following problem:

An opened iron-collapse with no content sets style fixed as "max-height: 0px", since the transition is from 0px->0px and thus _transitionEnd is not fired. When the iron-collapse receives content, it won't be displayed even thoug the iron-collapse is opened.

Fixes https://github.com/PolymerElements/iron-collapse/issues/64, fixes https://github.com/PolymerElements/iron-collapse/issues/65 (added by @valdrinkoshi)

### CODE REVIEW

**Comment by beta-vulgaris** [comment]
I signed it!

**Comment by beta-vulgaris** [comment]
I signed it!

**Comment by valdrinkoshi** [comment]
Hi @beta-vulgaris, thanks for the contribution! 
please follow our [contribution guidelines](https://github.com/PolymerElements/iron-collapse/blob/master/CONTRIBUTING.md) and open an issue exposing the problem (with a test case/jsbin), link the PR to that issue and provide unit tests to avoid any future regression :)

**Comment by beta-vulgaris** [comment]
PolymerElements/iron-collapse#64

**Comment by beta-vulgaris** [comment]
hi @valdrinkoshi ,
I added the test... will keep the guidelines in mind for future pull requests!
cheers

File: iron-collapse.html [hunk 1]
```diff
@@ -216,6 +216,9 @@
       // Set the final size.
       if (size === 'auto') {
         this.style[this._dimensionMax] = '';
+      // call _transitionEnd if no transition is performed, styles need to be set
+      } else if (size === startSize) {
+        this._transitionEnd();
       } else {
         this.style[this._dimensionMax] = size;
       }
```
Comment (valdrinkoshi) [inline iron-collapse.html]: I'd suggest we take care of calling `_transitionEnd` when needed in `updateSize`:
```javascript
this._updateTransition(false);
var willAnimate = animated && !this.noAnimation && this._isDisplayed;
// If we can animate, must do some prep work.
if (willAnimate) {
  var startSize = this._calcSize();
  // For `auto` we must calculate what is the final size for the animation.
  // After the transition is done, _transitionEnd will set the size back to `auto`.
  if (size === 'auto') {
    this.style[this._dimensionMax] = '';
    size = this._calcSize();
  }
  // Go to startSize without animation.
  this.style[this._dimensionMax] = startSize;
  // Force layout to ensure transition will go. Set scrollTop to itself
  // so that compilers won't remove it.
  this.scrollTop = this.scrollTop;
  // Enable animation.
  this._updateTransition(true);
  // If final size is the same as startSize it will not animate.
  willAnimate = (size !== startSize);
}
// Set the final size.
this.style[this._dimensionMax] = (size === 'auto') ? '' : size;
// If it won't animate, call transitionEnd to set correct classes.
if (!willAnimate) {
  this._transitionEnd();
}
```
This will allow you to remove the check for `noAnimation` in `_openedChanged` https://github.com/PolymerElements/iron-collapse/pull/63/files#diff-9c9f96df8dd507160ccdd7e52f160ed3R262

Comment (beta-vulgaris) [inline iron-collapse.html]: Hi @valdrinkoshi ,
I disagree. My solution would call _transitionEnd when we want animations but when it was not called because of a transition from (0px->0px). You're solution would call _transitionEnd in ALL cases eventually. This would break the tests "enableTransition(false) disables animations" and changes the behaviour of the component outside of the corner case "animation of an empty iron-collapse".

Comment (valdrinkoshi) [inline iron-collapse.html]: That's because there is a bug in `_transitionEnd` which resets the `maxHeight` at the end of the transition for opened `iron-collapse` (see https://github.com/PolymerElements/iron-collapse/issues/65). I guess this requires to store the desired size in some private variables, and ensure that is set in `_transitionEnd`. Something like:

```javascript
updateSize: function(size, animated) {
  // Consider 'auto' as '', to take full size.
  size = size === 'auto' ? '' : size;
  // No change!
  if (this._desiredSize === size) {
    return;
  }
  this._desiredSize = size;
  this._updateTransition(false);
  // all the `willAnimate` things ...
},

// ...

_transitionEnd: function() {
  this.style[this._dimensionMax] = this._desiredSize;
  // [...]
```

I believe it's the responsibility of `updateSize` to call `_transitionEnd` if there will be no animations. `_openedChanged` shouldn't do that. WDYT?

Comment (beta-vulgaris) [inline iron-collapse.html]: Sounds good!
I just committed the changes, including tests for external calls to updateSize.

File: iron-collapse.html [hunk 2]
```diff
@@ -127,7 +127,15 @@
...
       },
-
+	  
+	  /**
+       * stores the desired size of the collapse body
+       *
+	   */
+	  _desiredSize: {
+		type: String,
+		value: ""
```
Comment (valdrinkoshi) [inline iron-collapse.html]: seems there are some problems with the indentation in the file, also could you use single quotes in here?

**Comment by valdrinkoshi** [comment]
@beta-vulgaris thanks for the update! I left a small comment regarding formatting/indentation, LGTM otherwise 👍

File: test/basic.html [hunk 3]
```diff
@@ -35,6 +35,14 @@
         </iron-collapse>
       </template>
     </test-fixture>
+	
+	
+	<test-fixture id="test-empty">
```
Comment (valdrinkoshi) [inline test/basic.html]: indentation issues on this file as well

File: test/flex.html [hunk 4]
```diff
@@ -134,6 +134,15 @@
           assert.equal(collapse.style.height, '');
           assert.equal(collapse.style.transitionProperty, 'max-width');
         });
+		
+		test('change size with updateSize', function(done) {
+			collapse.addEventListener('transitionend', function() {
```
Comment (valdrinkoshi) [inline test/flex.html]: indentation issues here too

File: test/horizontal.html [hunk 5]
```diff
@@ -79,7 +79,19 @@
           // Size should be immediately set.
           assert.equal(collapse.style.maxWidth, '0px');
         });
+		
+		test('change size with updateSize', function(done) {
+			collapse.addEventListener('transitionend', function() {
```
Comment (valdrinkoshi) [inline test/horizontal.html]: indentation issues here too

File: iron-collapse.html [hunk 6]
```diff
@@ -127,7 +127,15 @@
       noAnimation: {
         type: Boolean
       },
-
+	  
+	  /**
+       * stores the desired size of the collapse body
+       *
```
Comment (valdrinkoshi) [inline iron-collapse.html]: Could you add the `@private` in here?

File: iron-collapse.html [hunk 7]
```diff
@@ -256,14 +270,11 @@
       if (this.opened) {
         this.focus();
       }
-      if (this.noAnimation) {
-        this._transitionEnd();
-      }
     },
 
     _transitionEnd: function() {
       if (this.opened) {
```
Comment (valdrinkoshi) [inline iron-collapse.html]: this `if` can be removed

File: iron-collapse.html [hunk 8]
```diff
@@ -128,6 +128,14 @@
         type: Boolean
       },
 
+      /**
+       * stores the desired size of the collapse body
+       * @private
+       */
+      _desiredSize: {
+        type: String,
+        value: ""
```
Comment (valdrinkoshi) [inline iron-collapse.html]: nit: could you use single quotes for the value? `value: ''`

File: iron-collapse.html [hunk 9]
```diff
@@ -128,6 +128,14 @@
         type: Boolean
       },
 
+      /**
+       * stores the desired size of the collapse body
```
Comment (valdrinkoshi) [inline iron-collapse.html]: nit: can you capitalize first letter and end the sentence with a `.` ? (this will show in the docs on elements.polymer-project.org/elements/iron-collapse)

File: iron-collapse.html [hunk 10]
```diff
@@ -120,14 +120,22 @@
       },
 
       /**
-       * Set noAnimation to true to disable animations
+       * Set noAnimation to true to disable animations.
```
Comment (valdrinkoshi) [inline iron-collapse.html]: 💯

File: iron-collapse.html [hunk 11]
```diff
@@ -120,14 +120,22 @@
...
        *
        * @attribute noAnimation
        */
       noAnimation: {
         type: Boolean
       },
 
+      /**
+       * Stores the desired size of the collapse body.
```
Comment (valdrinkoshi) [inline iron-collapse.html]: 📖

**Comment by valdrinkoshi** [comment]
LGTM 👍 great job @beta-vulgaris, thanks!
merging this as soon as tests are green ;)
